libgalois/include/katana/PropertyGraph.h

```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
     return edge_properties()->schema();
   }
 
-  /// \returns the number of node types
-  size_t GetNodeTypesNum() const {
-    return node_type_name_to_type_set_ids_.size();
+  /// \returns the number of node atomic types
+  size_t GetNodeAtomicTypesNum() const {
+    return node_type_name_to_entity_type_ids_.size();
   }
 
-  /// \returns the number of edge types
-  size_t GetEdgeTypesNum() const {
-    return edge_type_name_to_type_set_ids_.size();
+  /// \returns the number of edge atomic types
+  size_t GetEdgeAtomicTypesNum() const {
+    return edge_type_name_to_entity_type_ids_.size();
   }
 
-  /// \returns true if a node type with @param name exists
-  /// NB: no node may have this type
+  /// \returns the number of node EntityTypeIDs (including kUnknownType)
+  size_t GetNodeEntityTypesNum() const {
+    return node_entity_type_id_to_type_names_.size();
+  }
+
+  /// \returns the number of edge EntityTypeIDs (including kUnknownType)
+  size_t GetEdgeEntityTypesNum() const {
+    return edge_entity_type_id_to_type_names_.size();
+  }
+
+  /// \returns true iff a node atomic type with @param name exists
+  /// NB: no node may have a type that intersects with this atomic type
   /// TODO(roshan) build an index for the number of nodes with the type
-  bool HasNodeType(const std::string& name) const {
-    return node_type_name_to_type_set_ids_.count(name) == 1;
+  bool HasAtomicNodeType(const std::string& name) const {
+    return node_type_name_to_entity_type_ids_.count(name) == 1;
   }
 
-  /// \returns true if an edge type with @param name exists
-  /// NB: no edge may have this type
+  /// \returns true iff an edge atomic type with @param name exists
+  /// NB: no edge may have a type that intersects with this atomic type
   /// TODO(roshan) build an index for the number of edges with the type
-  bool HasEdgeType(const std::string& name) const {
-    return edge_type_name_to_type_set_ids_.count(name) == 1;
+  bool HasAtomicEdgeType(const std::string& name) const {
+    return edge_type_name_to_entity_type_ids_.count(name) == 1;
   }
 
-  /// \returns the set of node TypeSetIDs that contain
-  /// the node type with @param name
+  /// \returns true iff a node type with ID @param node_entity_type_id exists
+  /// NB: even if it exists, it may not be the most specific type for any node
+  /// (returns true for kUnknownType)
+  bool HasNodeEntityType(EntityTypeID node_entity_type_id) const {
+    return node_entity_type_id < node_entity_type_id_to_type_names_.size();
```

Comment:
Shouldn't there be a stronger check here, e.g., `&& !node_entity_type_id_to_type_names_[id].empty()` ?